### PR TITLE
impl(GCS+gRPC): more options for metrics exporter

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13998

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14238)

### CODE REVIEW

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (alevenberg) [inline google/cloud/storage/internal/grpc/metrics_exporter_options.cc:55]: Why are we creating a uuid?Is there something from the service we can use instead?

**Review by alevenberg — APPROVED** [review]
I don't have a ton of context, so this is more of a rubber stamp. You might want to wait for a second review.

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_options.cc:63]: nit: is the following more legible?

```cc
auto& labels = *monitored_reosurce.mutable_labels();
labels["project_id"] = project.project_id();
// etc.
```

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_options.cc:77]: Fine for now, but if this option defaulter function becomes more complex, consider factoring out some of the implementation details.

```cc
{
  auto monitored_resource = MakeMonitoredResource(...);
  auto endpoint = CalculateEndpointOption(...);
  return options.set<O1>(...).set<O2>(...).set<O3>(...);
}
```

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc:44]: nit: This is not actually empty. This is `opentelemetry::sdk::resource::Resource::GetDefault()` which has things like the otel library version.

Not that it matters, but I think you want: `opentelemetry::sdk::resource::Resource::GetEmpty()`

https://github.com/open-telemetry/opentelemetry-cpp/blob/ce14bf6520a059ff0bca81df45603722d965ad7c/sdk/include/opentelemetry/sdk/resource/resource.h#L56

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc:158]: Prior art says this can be simplified: https://github.com/googleapis/google-cloud-cpp/blob/dc92806e6190a235a23d64d1d7bbb0d4d5425b81/generator/integration_tests/tests/request_id_connection_impl_test.cc#L64-L68

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc:158]: That prior art depends on complexifying other parts of the code:

https://github.com/googleapis/google-cloud-cpp/blob/dc92806e6190a235a23d64d1d7bbb0d4d5425b81/generator/integration_tests/tests/request_id_connection_impl_test.cc#L90-L92

Comment (dbolduc) [reply google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc:158]: (I am not asking for changes, just FYI)

There is also prior art on avoiding `::testing::*Regex`

https://github.com/googleapis/google-cloud-cpp/blob/dc92806e6190a235a23d64d1d7bbb0d4d5425b81/google/cloud/testing_util/validate_metadata.cc#L102-L113

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options.cc]: Done.

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options.cc]: Thanks, it was not as terrible with `emplace()`, but your version is cleaner than even that.

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options.cc:61]: It is a required in the "spec", but it deserves a comment here. Thanks.

File: google/cloud/storage/internal/grpc/metrics_exporter_options.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options.cc:77]: Ack.

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc:44]: Changed to `TestResource()`.

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (coryan) [reply google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc]: Done

**Review by coryan — COMMENTED** [review]
Address review comments.  There was a minor change to the name of the monitored resource type in the last 24 hours.

File: google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc
Comment (dbolduc) [reply google/cloud/storage/internal/grpc/metrics_exporter_options_test.cc]: Nice. Thanks.

**Pull request merged**
